src/main/scala/uk/gov/nationalarchives/consignmentexport/ChecksumValidator.scala

```diff
@@ -0,0 +1,24 @@
+package uk.gov.nationalarchives.consignmentexport
+
+import java.util.UUID
+
+import cats.effect.IO
+import gov.loc.repository.bagit.domain.Bag
+import io.chrisdavenport.log4cats.SelfAwareStructuredLogger
+import uk.gov.nationalarchives.consignmentexport.Validator.ValidatedFileMetadata
+
+import scala.jdk.CollectionConverters._
+
+class ChecksumValidator()(implicit val logger: SelfAwareStructuredLogger[IO]) {
+
+  def validateFileChecksums(bag: Bag, validatedFileMetadata: List[ValidatedFileMetadata]): IO[List[UUID]] = {
+    val bagitGeneratedChecksums = bag.getPayLoadManifests.asScala.head.getFileToChecksumMap.values
+
+    IO(validatedFileMetadata.filterNot(fm => bagitGeneratedChecksums.contains(fm.clientSideChecksum)).map(_.fileId))
```

Comment:
Is this `IO` necessary? Did you find out whether it reads the checksums from memory?

Reply:
It should be reading from memory as its just a java object that's been created then passed into the function.